src/function/scalar/string/reverse.cpp

```diff
@@ -0,0 +1,80 @@
+#include "duckdb/function/scalar/string_functions.hpp"
+
+#include "duckdb/common/exception.hpp"
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+
+#include <string.h>
+
+using namespace std;
+
+namespace duckdb {
+
+typedef void (*str_function)(const char *input, char *output);
+
+static void strreverse(const char *input, char *output) {
+
+	size_t n = strlen(input); // the size of the string in bytes
+	size_t bytes = 0;
+
+	output[n] = 0;
+
+	while (*input) {
+		if (!(*input & 0x80)) { // !*input & 0b10000000
+			bytes = 1;
+		} else if ((*input & 0xe0) == 0xc0) { // (*input & 0b1110_0000 == 0b1100_0000)
```

Comment:
This is trying to compute UTF char length, right? There is code in value.cpp (IsUTF8String) that does something similar, perhaps those functions can be either merged or the UTFCharLen function should be moved there.

Reply:
Yes, it appears very similar.
My proposed back-of-the-envelope solution is to have two additional functions:
UTF8charlen(char c); that returns 1 2 3 or 4 and -1 for error, and UTF8isvalid(char* c); that returns true or false if the present memory position is a valid utf8 string, using UTF8charlen inside.

Reply:
is a valid UTF8 code point, not string. The string check should still be done by the IsUTF8String function, using the helper functions.

Reply:
This is the solution I came up with:

```
/**
 * returns:
 *  0 for valid, [1, 4] for valid and a length, -1 for invalid code point
 */
int Value::UTF8CodeLength(const char *s) {
	int c = *s;

	if (s == nullptr) {
		return 0;
	}
	if (*s == '\200' && s[1] == '\0') {
		return 0; /* str_nil */
	}

	if ((c & 0x80) == 0)
		return 1; 	// valid with length 1
	if ((*s++ & 0xC0) != 0x80)
		return -1;
	if ((c & 0xE0) == 0xC0)
		return 2;	// valid with length 2
	if ((*s++ & 0xC0) != 0x80)
		return -1;
	if ((c & 0xF0) == 0xE0)
		return 3;	// valid with length 3
	if ((*s++ & 0xC0) != 0x80)
		return -1;
	if ((c & 0xF8) == 0xF0)
		return 4;	// valid with length 4
	return -1;
}

// adapted from MonetDB's str.c
bool Value::IsUTF8String(const char *s) {
	int c;
	int len;

	len = UTF8CodeLength(s);

	if (len == 0) {
		return true;
	}
	
	while ((len = UTF8CodeLength(s)) > 0) {
		s += len;
	}
	return len > -1 ? true : false;
}
```

The AddressSanitizer complains about the code though, so I cannot submit it.